File: lib/state.rb
require 'ast/sequence'
require 'ast/variable'

class State
  def initialize(values = {})
    @values = values
  end

  attr_reader :values

  def assign_values(more_values)
    self.class.new(values.merge(more_values) { |key| raise key })
  end

  def value_of(key)
    if values.has_key?(key)
      value_of values.fetch(key)
    elsif key.is_a?(AST::Sequence)
      AST::Sequence.new(key.expressions.map(&method(:value_of)))
    else
      key
    end
  end

  def unify(a, b)
    a, b = value_of(a), value_of(b)

    if a == b
      self
    elsif a.is_a?(AST::Variable)
      assign_values a => b
    elsif b.is_a?(AST::Variable)
      assign_values b => a
    elsif a.is_a?(AST::Sequence) && b.is_a?(AST::Sequence)
      if a.expressions.length == b.expressions.length
        [a, b].map(&:expressions).transpose.inject(self) do |state, (a, b)|
          state && state.unify(a, b)
        end
      end
    end
  end
end

File: src/state.js
import AST from './ast';
import Builder from './builder'
import { Map } from 'immutable';

export default class State {
  constructor(values = Map()) {
    this.values = values;
  }

  assignValue(key, value) {
    return new State(this.values.set(key, value));
  }

  valueOf(key) {
    if (this.values.has(key)) {
      return this.valueOf(this.values.get(key));
    } else if (key instanceof AST.Sequence) {
      return new Builder().buildSequence(key.expressions.map(this.valueOf, this));
    } else {
      return key;
    }
  }

  unify(a, b) {
    const [av, bv] = [a, b].map(this.valueOf, this);

    if (av.equals(bv)) {
      return this;
    } else if (av instanceof AST.Variable) {
      return this.assignValue(av, bv);
    } else if (bv instanceof AST.Variable) {
      return this.assignValue(bv, av);
    } else if (av instanceof AST.Sequence && bv instanceof AST.Sequence) {
      if (av.expressions.size === bv.expressions.size) {
        return av.expressions.zip(bv.expressions).
          reduce((state, [a, b]) => state && state.unify(a, b), this);
      }
    }
  }
};
